geode-core/src/distributedTest/java/org/apache/geode/cache30/CacheStatisticsPartitionedRegionDUnitTest.java

```diff
@@ -0,0 +1,577 @@
+    region.get(key);
+
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 1 == rStats.getMissCount());
+    assertEquals(0, rStats.getHitCount());
+    assertEquals(1, rStats.getMissCount());
+  }
+
+  /**
+   * Tests that the {@linkplain CacheStatistics#getLastAccessedTime last access time} and
+   * {@link CacheStatistics#getLastModifiedTime last modified time} are updated appropriately for a
+   * partitioned region with only one server.
+   */
+  @Test
+  public void testTimeStats() throws CacheException, InterruptedException {
+    final long ESTAT_RES = 100; // the resolution, in ms, of entry stats
+    String name = this.getUniqueName();
+    Object key = "KEY";
+    Object key2 = "KEY2";
+    Object value = "VALUE";
+    long before;
+    long after;
+    long oldBefore;
+    long oldAfter;
+
+    final int MSECS = 1000;
+
+    AttributesFactory factory = new AttributesFactory();
+    // factory.setScope(Scope.LOCAL);
+    factory.setStatisticsEnabled(true);
+
+    Region region = createPartitionedRegion(name, factory.create());
+    CacheStatistics rStats = region.getStatistics();
+
+    before = ((GemFireCacheImpl) getCache()).cacheTimeMillis();
+    // In order for the stats to have a value, an entry must be accessed first.
+    region.get(key);
+    after = ((GemFireCacheImpl) getCache()).cacheTimeMillis();
+
+    {
+      final long localBefore = before;
+      final long localAfter = after;
+      GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+          .until(() -> (localBefore <= rStats.getLastAccessedTime()
+              && localAfter >= rStats.getLastAccessedTime()));
+    }
+    assertInRange(before, after, rStats.getLastModifiedTime());
+
+    oldBefore = before;
+    oldAfter = after;
+    before = ((GemFireCacheImpl) getCache()).cacheTimeMillis();
+    region.get(key);
+    after = ((GemFireCacheImpl) getCache()).cacheTimeMillis();
+
+    {
+      final long localBefore = before;
+      final long localAfter = after;
+      GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+          .until(() -> localBefore <= rStats.getLastAccessedTime()
+              && localAfter >= rStats.getLastAccessedTime());
+    }
+    assertInRange(oldBefore, oldAfter, rStats.getLastModifiedTime());
+
+    before = ((GemFireCacheImpl) getCache()).cacheTimeMillis();
+    region.put(key, value);
+    CacheStatistics eStats = region.getEntry(key).getStatistics();
+    after = ((GemFireCacheImpl) getCache()).cacheTimeMillis();
+    {
+      final long localBefore = before;
+      final long localAfter = after;
+      GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+          .until(() -> localBefore <= rStats.getLastModifiedTime()
+              && localAfter >= rStats.getLastModifiedTime());
+    }
+    assertInRange(before, after, rStats.getLastAccessedTime());
+
+    assertInRange(before - ESTAT_RES, after + ESTAT_RES, eStats.getLastAccessedTime());
```

Comment:
For these entry stat ones we probably need to wait until it has changed by more than ESTAT_RES.

Reply:
Yes.